File: scripts/gen-icons.mjs
/**
 * Генерация PWA-иконок RIP.
 * Использование: node scripts/gen-icons.mjs
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { createCanvas, loadImage, registerFont } from 'canvas';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const root = path.join(__dirname, '..');
const out = path.join(root, 'public', 'icons');
fs.mkdirSync(out, { recursive: true });

// Georgia шрифт
registerFont('C:\\Windows\\Fonts\\georgia.ttf', { family: 'Georgia' });

function drawRip(canvas) {
  const ctx = canvas.getContext('2d');
  const S = canvas.width;
  const c = S / 2; // 256

  // 1. Чёрный фон со скруглением
  const r = Math.round(S * 112 / 512); // радиус скругления
  ctx.beginPath();
  ctx.roundRect(0, 0, S, S, r);
  ctx.fillStyle = '#000000';
  ctx.fill();

  // 2. Тень под надгробием
  ctx.beginPath();
  ctx.ellipse(c, S * 410 / 512, S * 150 / 512, S * 16 / 512, 0, 0, Math.PI * 2);
  ctx.fillStyle = '#141414';
  ctx.fill();

  // 3. Надгробье
  const t = S * 156 / 512; // left
  const b = S * 410 / 512; // bottom
  const top = S * 220 / 512; // top of arch
  const w = S * 200 / 512; // width (356-156)
  const cx = c; // center
  const aR = w / 2; // arch radius = 100
  ctx.beginPath();
  ctx.moveTo(t, b);
  ctx.lineTo(t, S * 220 / 512);
  // арка: 100px радиус, 180° от 0 до π
  ctx.arc(cx, S * 220 / 512, aR, Math.PI, 0, false);
  ctx.lineTo(t + w, b);
  ctx.closePath();
  ctx.fillStyle = '#2E2E2E';
  ctx.fill();

  // 4. Внутренняя рамка (offset 22px)
  const inset = S * 22 / 512;
  const t2 = t + inset;
  const top2 = S * 226 / 512; // 220 + 6
  const w2 = w - 2 * inset;
  const aR2 = w2 / 2;
  ctx.beginPath();
  ctx.moveTo(t2, b - inset);
  ctx.lineTo(t2, top2);
  ctx.arc(cx, top2, aR2, Math.PI, 0, false);
  ctx.lineTo(t2 + w2, b - inset);
  ctx.closePath();
  ctx.strokeStyle = '#4B4B4B';
  ctx.lineWidth = S * 4 / 512;
  ctx.stroke();

  // 5. Текст RIP
  const fontSize = Math.round(S * 76 / 512);
  ctx.font = `700 ${fontSize}px Georgia, serif`;
  ctx.textAlign = 'center';
  ctx.textBaseline = 'middle';
  ctx.fillStyle = '#C8C8C8';
  ctx.letterSpacing = '6px';
  // letterSpacing не поддерживается в canvas, применяем через char spacing
  const text = 'RIP';
  const firstChar = text[0];
  const charX = c - (text.length - 1) * 10; // 10px spacing / 2
  ctx.fillText(text, c, S * 286 / 512);
  // геометрия: чуть выше центра арки

  // 6. Линия земли
  ctx.beginPath();
  ctx.moveTo(S * 86 / 512, b);
  ctx.lineTo(S * 426 / 512, b);
  ctx.strokeStyle = '#1C1C1C';
  ctx.lineWidth = S * 8 / 512;
  ctx.lineCap = 'round';
  ctx.stroke();
}

// --- 512 ---
const cv512 = createCanvas(512, 512);
drawRip(cv512);
const buf512 = cv512.toBuffer('image/png');
fs.writeFileSync(path.join(out, 'icon-512.png'), buf512);
console.log('icon-512.png', buf512.length, 'bytes');

// --- 192 ---
const cv192 = createCanvas(192, 192);
drawRip(cv192);
const buf192 = cv192.toBuffer('image/png');
fs.writeFileSync(path.join(out, 'icon-192.png'), buf192);
console.log('icon-192.png', buf192.length, 'bytes');

// --- maskable 512 (с отступом 20%) ---
const cvMask = createCanvas(512, 512);
const ctxMask = cvMask.getContext('2d');
const pad = 512 * 0.1; // 10% отступ = 80% safe area
ctxMask.save();
ctxMask.translate(pad, pad);
ctxMask.scale(0.8, 0.8);
drawRip(cvMask);
ctxMask.restore();
const bufMask = cvMask.toBuffer('image/png');
fs.writeFileSync(path.join(out, 'icon-maskable-512.png'), bufMask);
console.log('icon-maskable-512.png', bufMask.length, 'bytes');

console.log('Готово');

File: scripts/gen_icons.py
#!/usr/bin/env python3
"""Генерация PWA-иконок RIP: 512, 192, maskable."""
import os
from PIL import Image, ImageDraw, ImageFont

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
OUT = os.path.join(ROOT, 'public', 'icons')
os.makedirs(OUT, exist_ok=True)

FONT = 'C:/Windows/Fonts/georgia.ttf'

def draw_rip(size):
    img = Image.new('RGB', (size, size), '#000000')
    d = ImageDraw.Draw(img)
    s = size / 512.0  # scale

    # скруглённый чёрный фон (радиус 112)
    mask = Image.new('L', (size, size), 0)
    md = ImageDraw.Draw(mask)
    md.rounded_rectangle([0, 0, size - 1, size - 1], radius=112 * s, fill=255)
    bg = Image.new('RGB', (size, size), '#000000')
    img = Image.composite(bg, img, mask)
    d = ImageDraw.Draw(img)

    # тень
    d.ellipse([256 * s - 150 * s, 410 * s - 16 * s, 256 * s + 150 * s, 410 * s + 16 * s], fill='#141414')

    # надгробье
    d.polygon([(156 * s, 410 * s), (156 * s, 220 * s)], fill=None)
    d.pieslice([156 * s, 120 * s, 356 * s, 320 * s], 180, 360, fill='#2E2E2E')
    d.rectangle([156 * s, 220 * s, 356 * s, 410 * s], fill='#2E2E2E')

    # внутренняя рамка (stroke)
    d.arc([178 * s, 148 * s, 334 * s, 304 * s], 180, 360, fill='#4B4B4B', width=max(2, int(4 * s)))
    d.line([(178 * s, 226 * s), (178 * s, 402 * s)], fill='#4B4B4B', width=max(2, int(4 * s)))
    d.line([(334 * s, 226 * s), (334 * s, 402 * s)], fill='#4B4B4B', width=max(2, int(4 * s)))
    d.line([(178 * s, 402 * s), (334 * s, 402 * s)], fill='#4B4B4B', width=max(2, int(4 * s)))

    # текст RIP
    font = ImageFont.truetype(FONT, int(76 * s))
    # letter-spacing 6: рисуем по буквам
    text = 'RIP'
    bbox = d.textbbox((0, 0), text, font=font)
    tw = bbox[2] - bbox[0]
    spacing = 6 * s * 2  # letter-spacing с обеих сторон каждой буквы кроме первой? просто добавим суммарно
    total_w = tw + 6 * s * 2  # 2 промежутка между 3 буквами
    x = 256 * s - total_w / 2
    for ch in text:
        cb = d.textbbox((0, 0), ch, font=font)
        d.text((x - cb[0], 286 * s - (cb[1] + cb[3]) / 2), ch, font=font, fill='#C8C8C8')
        chw = cb[2] - cb[0]
        x += chw + 6 * s

    # линия земли
    d.line([(86 * s, 410 * s), (426 * s, 410 * s)], fill='#1C1C1C', width=max(3, int(8 * s)))

    return img

img512 = draw_rip(512)
img512.save(os.path.join(OUT, 'icon-512.png'))
print('icon-512.png', os.path.getsize(os.path.join(OUT, 'icon-512.png')))

img192 = draw_rip(192)
img192.save(os.path.join(OUT, 'icon-192.png'))
print('icon-192.png', os.path.getsize(os.path.join(OUT, 'icon-192.png')))

# maskable: контент в 80% центра
pad = int(512 * 0.1)
canvas = Image.new('RGB', (512, 512), '#000000')
inner = draw_rip(512).resize((512 - pad * 2, 512 - pad * 2), Image.LANCZOS)
canvas.paste(inner, (pad, pad))
canvas.save(os.path.join(OUT, 'icon-maskable-512.png'))
print('icon-maskable-512.png', os.path.getsize(os.path.join(OUT, 'icon-maskable-512.png')))

print('Готово')
